### ISSUE TITLE
Discussion: Doest it make sense DeviceStorageDataSource.getAll ?

### ISSUE THREAD

**Issue by RubenVot** [mobilejazz/harmony-swift#11]
I feel a bit confused with the behaviour of DeviceStorageDataSource 

If I put two objects: 
```
DeviceStorageDatasource.put(key: "key1", object1<Type1>)
DeviceStorageDatasource.put(key: "key2", object2<Type1>)
```
And then 
```
DeviceStorageDataSource.getAll()
```
There is no way to get those two objects because DeviceStorageDataSource will try to find  one array maching the provided key in the query. But the provided key, because of the limitations of UserDefaults, can only reference one object

I think that getAll/putAll should be not allowed for DeviceDataStorageDataSource

**Comment by joselufo** [comment]
Good point Ruben. I think the implementations of DeviceStorageDataSource will change a bit depending on the platform that implements it as every platform has a different way to work. For example, in Android, we don't have a way to save an array in the SharePreferences and with the UserDefault, you can. 

I think if the UserDefault allows to store and fetch an array it makes sense to have implemented the getAll and putAll. 

What do you think?

**Comment by vilanovi** [comment]
Maybe here we need to think deeper:

Maybe the default implementation of a `dataSource.getAll()` should use the `AllObjectsQuery()`. This would then justify that your `getAll` would return all stored objects. However, note that this is not always possible, as in the `UserDefaults` case, you must know the keys in advance, otherwise you would be returning unkonwn objects stored in the user defaults (as for example, the languages used in the device and other system stored keys).

Otherwise, maybe the problem is that you expect the system to be "too smart". At the end, the `DeviceStorageDataSource` is quite primitive. You store something with a key, you can then retrieve that something with the same key. Nothing else.

What do you think then Rubén? how we should focus this?

**Comment by vilanovi** [comment]
Note that `UserDefauts`allow to store Arrays of objects. This could justify the existence of `getAll` and `putAll` methods.

**Comment by vilanovi** [comment]
Ok, this evening I was thinking about how to improve this. After dinner, I was eager to play a bit with the code and I've implemented for fun an approach where the writing to the `UserDefaults` happens in a more organized way.

The current implementation in Harmony Swift 1.0 of this class is a dumb get/put/delete to the UserDefaults storage. This actually leads to some weird behaviors, as you were describing Rubén, but also things like writing a value using a `DeviceStorageDataSource<Int>` and reading it using a `DeviceStorageDataSource<String>` (or any combination you want). 

Additionally, there is no way to add a default implementation for a `deleteAll(AllObjectsQuery())`, as inside the UserDefaults there are other values that coexist with the values inserted by a specific DataSource (and we don't know the keys to delete them).

Therefore, what I have done is modify the current implementation to `DeviceStorageDataSource` and give it 3 possible behaviors (note that all this is up for discussion) and a new initializer:
```swift
public init(_ userDefaults: UserDefaults = UserDefaults.standard, storageType: DeviceStorageType = .regular) 
```
where:
```swift
public enum DeviceStorageType {
    case regular
    case prefix(String)
    case rootKey(String)
}
```

##### Case `.regular`:
Default behavior. Accessing regular values by their key in `UserDefaults`.

##### Case `.prefix(let prefix)`:
Adding a prefix to the given userDefaults key. For example, a `Query` (implementing `KeyQuery`) with a key `myKey` would be stored to `UserDefaults` using the key `prefix.myKey`. 

This is done to prevent conflicts of two independents data sources writing on the same key name (each data source could have a different prefix).

Note this was implemented in the first version of Harmony and we can decide to deprecate it if we don't want it.

##### Case `.rootKey(let rootKey)`:
This is the new behavior. Mostly, the dataSource will create an empty dictionary and store it inside `UserDefaults` on the key `rootKey`. Then, every value stored and retrieved from the dataSource will be stored or fetched from that dictionary.

For example, if you store the value `$customValue` for the key `$customKey`, the `DeviceStorageDataSource<Type>` will save it in user defaults as:

```json
{ 
    "other_user_default_keys_1" : "some_values",
    "rootKey<Type>" :  {
        "$customKey": $customValue
     },
    "other_user_default_keys_2" : "some_values",
}
```

Note that the data source's type is also part of that root key, meaning multiple data sources of different type but with the same root key wouldn't conflict. 

Then, and as a free add-on, I could implement support for the `AllObjectsQuery` in`deleteAll` methods, deleting all objects for that key (actually, deleting the `rootKey<Type>` dictionary). We could even implement (I haven't done it yet) the `getAll` method as well (but here we need to be careful, as we could mix simple objects with array of objects within an array).

In any case, this behavior brings a more organized, self-contained, collision free (no keys conflicts between other `DeviceStorageDataSources`), and more powerful way of interacting with our `UserDefaults`. 

I am aware that this behavior has also a performance drawback, as every time is serializing and deserializing the whole dictionary. --> IDEA: We could otherwise add a key filtering on the `.prefix(String)` case and implement an "AllObjectsQuery" support in deleteAll and getAll.

##### In conclusion

I just wanted to share some deeper thoughts on how we can improve the `DeviceStorageDataSource` in swift, by doing a more sophisticated manipulation of `UserDefaults`. 

All code I've done is on my local machine. I'll be glad to share it if you want to give it a look. I don't expect to push this to production at this stage (or at least not yet).

Thanks and good night,

**Comment by RubenVot** [comment]
Good work @vilanovi ! 
Actually, I had been thinking about the same approach on Sunday when I found this problem. I was doing a utility for myself and I solved the problem using the same approach: adding a prefix. (However, I did a fast patch, not nice code as you). 

My conclusions were very similar also regarding to serialize/deserialize. Every time you modify an object you need to store everything, but I´m not sure this might be a problem taking into account that most of the operations will happen in a background executor. 

Also, the developer should read the docs and understand that this is not the proper place to store 1000 values. 

For me, it´s a bad solution only if the methods don´t do what they are supposed to do. A lack of performance is just a characteristic of the storage and whoever uses it should be aware about it.

So. I say yes!

**Comment by RubenVot** [comment]
@vilanovi are you planning to submit your PR soon? I would like to work on this implementation as I need it for a PoC. Or maybe you can open a branch and I can collaborate somehow

**Comment by vilanovi** [comment]
Pushed to `feature/devicestoragedatasource_improvements`

**Comment by RubenVot** [comment]
Nice, I just moved to my ToDo :)

**Comment by vilanovi** [comment]
@RubenVot any progress so far? I need this for a feature in PeerWell.

**Comment by RubenVot** [comment]
Uhmm I was using this in a personal project which I work in during the weekends and I couldn't do it last week because I was traveling. I would like to resume this weekend. But if you need it asap I'm afraid I won't be able to help :(

**Comment by vilanovi** [comment]
The new DeviceStatusDataSource and InMemoryDataSource have been released. You can now query AllObjectsQuery to retrieve them all or delete them all.
